src/eth-signer/src/ic_sign.rs

```diff
@@ -0,0 +1,228 @@
+}
+
+pub struct IcSigner;
+
+impl IcSigner {
+    /// Signs the transaction using `ManagementCanister::sign_with_ecdsa()` call.
+    pub async fn sign_transaction(
+        &self,
+        tx: &TypedTransaction,
+        key_id: SigningKeyId,
+        derivation_path: DerivationPath,
+    ) -> Result<Signature, IcSignerError> {
+        let hash = tx.sighash();
+        let hash_bytes = hash.as_fixed_bytes();
+
+        let request = SignWithECDSAArgs {
+            key_id: EcdsaKeyId {
+                curve: EcdsaCurve::Secp256k1,
+                name: key_id.to_string(),
+            },
+            message_hash: *hash_bytes,
+            derivation_path,
+        };
+        let signature_data = virtual_canister_call!(
+            Principal::management_canister(),
+            "sign_with_ecdsa",
+            (request,),
+            ic_exports::ic_ic00_types::SignWithECDSAReply
+        )
+        .await
+        .map_err(|(code, msg)| IcSignerError::SigningFailed(code, msg))?
+        .signature;
+
+        let r = ethers_core::types::U256::from_big_endian(&signature_data[0..32]);
+        let s = ethers_core::types::U256::from_big_endian(&signature_data[32..64]);
+
+        // IC doesn't support recovery id signature parameter, so set it manually.
+        // Details: https://eips.ethereum.org/EIPS/eip-155.
+        let v = match &tx {
+            TypedTransaction::Legacy(_) => tx.chain_id().unwrap_or_default().as_u64() * 2 + 35u64,
+            _ => 0,
+        };
+        let mut signature = Signature { v, r, s };
+
+        // Recovery id value may be increased by one, depending on internal
+        // signing parameter we don't know.
+        // The only thing we can do: try to recover address and, if failed,
+        // assume that recovery id should be increased.
+        let recovered = signature.recover(hash.0)?;
+        let tx_from = tx.from().ok_or(IcSignerError::FromAdderessNotPresent)?;
+        if &recovered != tx_from {
+            signature.v += 1;
+        };
+
+        Ok(signature)
+    }
+
+    /// Returns public key for current canister from IC.
+    pub async fn public_key(
+        &self,
+        key_id: SigningKeyId,
+        derivation_path: DerivationPath,
+    ) -> Result<Vec<u8>, IcSignerError> {
+        let request = ECDSAPublicKeyArgs {
+            canister_id: None,
+            derivation_path,
+            key_id: EcdsaKeyId {
+                curve: EcdsaCurve::Secp256k1,
+                name: key_id.to_string(),
+            },
+        };
+        virtual_canister_call!(
+            Principal::management_canister(),
+            "ecdsa_public_key",
+            (request,),
+            ECDSAPublicKeyResponse
+        )
+        .await
+        .map_err(|(code, msg)| IcSignerError::SigningFailed(code, msg))
+        .map(|response| response.public_key)
+    }
+
+    /// Convert public key to ethereum address.
+    pub fn pubkey_to_address(&self, pubkey: &[u8]) -> Result<H160, IcSignerError> {
```

Comment:
There is a test which checks correctness of all three functions in `IcSigner` using virtual responder and tx signing library instead of the actual IC signing. 
These methods contain IC-specific signing logic (e.g. compressed sec1 key encoding, signature recovery_id calculation, etc...), so I think we can't move this logic to some generic library.